Purchases/Public/RCPurchases.m

```diff
@@ -680,19 +694,19 @@ - (void)cachePurchaserInfo:(RCPurchaserInfo *)info forAppUserID:(NSString *)appU
     }
 }
 
-- (void)updateCaches {
-    [self updateCachesWithCompletionBlock:nil];
+- (void)updateAllCaches {
+    [self updateAllCachesWithCompletionBlock:nil];
 }
 
-- (void)updateCachesWithCompletionBlock:(nullable RCReceivePurchaserInfoBlock)completion
+- (void)updateAllCachesWithCompletionBlock:(nullable RCReceivePurchaserInfoBlock)completion
 {
-    [self.deviceCache resetCachesTimestamp];
-    [self updatePurchaserInfoCache:completion];
+    [self fetchAndCachePurchaserInfoWithCompletion:completion];
     [self updateOfferingsCache:nil];
 }
 
-- (void)updatePurchaserInfoCache:(nullable RCReceivePurchaserInfoBlock)completion
+- (void)fetchAndCachePurchaserInfoWithCompletion:(nullable RCReceivePurchaserInfoBlock)completion
 {
+    [self.deviceCache setPurchaserInfoCacheTimestampToNow];
```

Comment:
I believe this is the current behavior too, but why are we even setting this here? The backend call can fail 🤔 Wouldn't it be easier to just set the timestamp and the caches only if it's successful?

Reply:
I believe this is for thread safety, to ensure that if another caller comes in, they don't start a new backend request. 
There might be unintended side effects, though - like what happens if another caller comes in, sees that purchaser info was just cached, and tries to use it, but there's nothing there?

Reply:
note that if the backend call fails, this sets the timestamp to null so it can get reattempted

Reply:
synchronizing might be the best way around this, though.

Reply:
I don't want to go super deep into refactoring the behavior here in this PR, because there are already quite a few changes here. 
Maybe in a follow-up PR?

Reply:
and looking at this again, I think there's two behaviors: 
- if you have info cached, and call `purchaserInfoWithCompletionBlock` on two threads
  - first one returns old info and triggers a re-fetch
  - second returns old info but sees the re-fetch and doesn't make another request
- if you *don't have info cached*, and call `purchaserInfoWithCompletionBlock` on two threads
  - both re-fetch, because since there's nothing in cache, they decide to fetch. they only call completion when they get info. 

This should probably be fixed. But I still think it might be worth doing so on another PR, since I don't want this one to loose focus and cause new bugs. @vegaro thoughts?

Reply:
> both re-fetch, because since there's nothing in cache, they decide to fetch.

But when the second fetch comes, it would there is another call going on for that same call and wouldn't trigger a second call right? You can see that in `getSubscriberDataWithAppUserID` in `RCBackend` in:
```
    if ([self addCallback:completion forKey:path]) {
```

Something that I have seen happening before for offerings (and possible purchaser info) is:
1. SDK sets up and triggers a call to fetch the offerings but the offerings http call takes some time to complete
1. The app triggers a `getOfferings` right after setting up the SDK. Since the first call hasn't completed the cache is empty so it will start triggering the fetch of the offerings from the backend. 
1. The first call comes back before the `getOfferings` has checked if there is another call going on to fetch offerings (before it hits line 341 in RCBackend). So by the time it checks, there won't be any call going on and it will refetch again.

I believe all these we are talking are edge cases between the cache and the http calls. We should probably rethink them and open a future PR.

Reply:
you're right, I'd missed the behavior in RCBackend. 
I agree that this is worth putting some thought into and cleaning up. It's pretty confusing as it stands right now